packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -295,22 +295,43 @@ class RenderEditable extends RenderBox {
   ValueListenable<bool> get selectionEndInViewport => _selectionEndInViewport;
   final ValueNotifier<bool> _selectionEndInViewport = ValueNotifier<bool>(true);
 
+  // Returns a bool indicating if the offset is within factor of being contained
+  // within the rect.
+  static bool _containsFuzzy(Rect rect, Offset offset, [double factor = 0.1]) {
+    return rect.contains(offset)
+      || rect.contains(offset + Offset(factor, factor))
+      || rect.contains(offset + Offset(factor, -factor))
+      || rect.contains(offset + Offset(-factor, factor))
+      || rect.contains(offset + Offset(-factor, -factor));
+  }
+
   void _updateSelectionExtentsVisibility(Offset effectiveOffset) {
     final Rect visibleRegion = Offset.zero & size;
 
     final Offset startOffset = _textPainter.getOffsetForCaret(
       TextPosition(offset: _selection.start, affinity: _selection.affinity),
       Rect.zero,
     );
-
-    _selectionStartInViewport.value = visibleRegion.contains(startOffset + effectiveOffset);
+    final Offset offset = startOffset + effectiveOffset;
+    // Check if the caret is visible with an approximation because of an error
+    // in the engine that always reports that the caret extends above the input
+    // by a small amount. See paragraph.cc's implementation of GetRectsForRange.
```

Comment:
@GaryQian Is there anything else more specific that I can refer people to for the slightly off engine bug that I asked you about on Friday?

By the way, despite all of the tests that I added to this, I was never able to reproduce that engine bug in the tests.  Any idea how I might do that if it's possible?

Reply:
There's a few places we do height rounding for text: https://github.com/flutter/engine/blob/master/third_party/txt/src/txt/paragraph.cc#L921
and https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/painting/text_painter.dart#L294 come to mind. The combination of these produces integer height values.

Is this the bug you are asking about? Maybe because we are rounding in both the engine and the framework, one of those rounds is eating up your "slightly off".

Reply:
GetRectsForRange produces the box by using the rounded baseline, but then adds on the unrounded ascent and descent, which may cause inconsistencies between the reported height and the getRectsForRange positions.